src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -0,0 +1,557 @@
+    void ActionMap::AddAction(const Model::Command& cmd)
+    {
+        // _Never_ add null to the ActionMap
+        if (!cmd)
+        {
+            return;
+        }
+
+        // Handle nested commands
+        const auto cmdImpl{ get_self<Command>(cmd) };
+        if (cmdImpl->IsNestedCommand())
+        {
+            // But check if it actually has a name to bind to first
+            const auto name{ cmd.Name() };
+            if (!name.empty())
+            {
+                _NestedCommands.Insert(name, cmd);
+            }
+            return;
+        }
+
+        // General Case:
+        //  Add the new command to the KeyMap.
+        //  This map directs you to an entry in the ActionMap.
+
+        // Removing Actions from the Command Palette:
+        //  cmd.Name and cmd.Action have a one-to-one relationship.
+        //  If cmd.Name is empty, we must retrieve the old name and remove it.
+
+        // Removing Key Bindings:
+        //  cmd.Keys and cmd.Action have a many-to-one relationship.
+        //  If cmd.Keys is empty, we don't care.
+        //  If action is "unbound"/"invalid", you're explicitly unbinding the provided cmd.keys.
+        //  NOTE: If we're unbinding a command from a different layer, we must use ConsolidatedActions
+        //         to keep track of what key mappings are still valid.
+
+        // _TryUpdateActionMap may update oldCmd and consolidatedCmd
+        Model::Command oldCmd{ nullptr };
+        Model::Command consolidatedCmd{ nullptr };
+        _TryUpdateActionMap(cmd, oldCmd, consolidatedCmd);
+
+        _TryUpdateName(cmd, oldCmd, consolidatedCmd);
+        _TryUpdateKeyChord(cmd, oldCmd, consolidatedCmd);
+    }
+
+    // Method Description:
+    // - Try to add the new command to _ActionMap.
+    // - If the command was added previously in this layer, populate oldCmd.
+    // - If the command was added previously in another layer, populate consolidatedCmd.
+    // Arguments:
+    // - cmd: the action we're trying to register
+    // - oldCmd: the action found in _ActionMap, if one already exists
+    // - consolidatedAction: the action found in a parent layer, if one already exists
+    void ActionMap::_TryUpdateActionMap(const Model::Command& cmd, Model::Command& oldCmd, Model::Command& consolidatedCmd)
+    {
+        const auto actionID{ ActionHash{}(cmd.ActionAndArgs()) };
+
+        const auto& actionPair{ _ActionMap.find(actionID) };
+        if (actionPair != _ActionMap.end())
+        {
+            // We're adding an action that already exists in our layer.
+            // Record it so that we update it with any new information.
+            oldCmd = actionPair->second;
+        }
+        else
+        {
+            // add this action in for the first time
+            _ActionMap.emplace(actionID, cmd);
+        }
+
+        // Now check if this action was introduced in another layer.
+        const auto& consolidatedActionPair{ _ConsolidatedActions.find(actionID) };

```

Comment:
it took me 300 lines to understand that "Consolidated" was the "commands that we modified which came from a lower layer" that we talked about

Reply:
hahaha. I hate every little bit of it. The current name is a bit too long for my liking, but I needed a way to distinguish it from the one from `_ActionMap`. :/ _Very_ open to suggestions for a different name.

For the record, the original name was `NonSerializableAction` hahaha